### PR TITLE
Remove predefined resources vector in ResourceRequest

### PR DESCRIPTION
**Pull request by raulchen**
## Why are these changes needed?

"ResourceRequest" now uses 2 containers: a vector for predefined resources, and a map for custom resources. 
This was intended to be a perf optimization. However, in practice, this makes the code more complex, and, moreover, prevents optimizations for some methods (e.g., "ResourceIds", "Size").

This PR removes the vector and makes ResourceRequest use only one map for all resources. Also, "ResourceIds" now returns a "boost:range" to allow iterating resource IDs without having to construct temporary sets. 

microbenchmark shows a slight perf improvement.
last nightly: `placement group create/removal per second 837.76 +- 16.68`.
this PR: `placement group create/removal per second 895.76 +- 16.99`.

## Related issue number

#22850

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by raulchen** [comment]
Full microbenchmark result. Left is this PR, right is latest nightly.
image

**Comment by scv119** [comment]
nice!

File: src/ray/raylet/scheduling/cluster_resource_data.h [hunk 1]
```diff
@@ -61,97 +66,56 @@ class ResourceRequest {
...
-    }
-    for (auto &entry : custom_resources_) {
-      res.insert(ResourceID(entry.first));
-    }
-    return res;
-  }
+  bool IsEmpty() const { return resources_.empty(); }
+
+  /// Return a boost::range object that can be used as an iterator of the resource IDs.
+  ResourceIdIterator ResourceIds() const { return boost::adaptors::keys(resources_); }
```
Comment (scv119) [inline src/ray/raylet/scheduling/cluster_resource_data.h:112]: what happens to the iterator if the resources_ is changed?

**Review by scv119 — COMMENTED** [review]
LGTM! @jjyao @rkooo567 could you take a look?

**Review by scv119 — APPROVED** [review]

File: src/ray/raylet/scheduling/cluster_resource_data.h (hunk 1 above)
Comment (raulchen) [reply src/ray/raylet/scheduling/cluster_resource_data.h:112]: Good question. I could find the documentation about this. 
But I did 2 tests. They work well.
```
Test 1:
  r1 = ResourceRequest({{cpu_id, -1}, {custom_id1, 2}, {custom_id2, -2}});
  for (auto &resource_id : r1.ResourceIds()) {
    r1.Set(resource_id, 0);
    RAY_LOG(INFO) << resource_id;
  }
  
Prints:
[2022-03-31 08:32:47,385 I 185337 185337] resource_request_test.cc:158: CPU
[2022-03-31 08:32:47,385 I 185337 185337] resource_request_test.cc:158: custom2
[2022-03-31 08:32:47,385 I 185337 185337] resource_request_test.cc:158: custom1

Test 2:
  r1 = ResourceRequest({{cpu_id, -1}, {custom_id1, 2}, {custom_id2, -2}});
  for (auto &resource_id : r1.ResourceIds()) {
    r1.Clear();
    RAY_LOG(INFO) << resource_id;
  }

Prints:
[2022-03-31 08:23:29,366 I 180899 180899] resource_request_test.cc:158: custom2


```

File: src/ray/raylet/scheduling/cluster_resource_data.h [hunk 2]
```diff
@@ -363,36 +286,28 @@ class TaskResourceInstances {
...
   TaskResourceInstances &Set(const ResourceID resource_id,
                              const std::vector<FixedPoint> &instances) {
     if (instances.size() == 0) {
       Remove(resource_id);
     } else {
-      auto ptr = GetPointer(resource_id);
-      if (ptr != nullptr) {
-        *ptr = instances;
-      } else {
-        custom_resources_.emplace(resource_id.ToInt(), instances);
```
Comment (raulchen) [inline src/ray/raylet/scheduling/cluster_resource_data.h:387]: @scv119 I just found a bug introduced by last PR. 
The impact is that , for custom resources, only the first Set will work.
If you see any related issue on the nightly tests, this PR can fix the issue.

**Review by jjyao — APPROVED** [review]

**Pull request merged**